File: srcs/juloo.keyboard2/MaxHeightListView.java
package juloo.keyboard2;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.widget.ListView;

/**
 * A scrollable ListView that limits its maximum height.
 * Used for pinned clipboard items - shows max 2 items with internal scrolling.
 */
public class MaxHeightListView extends ListView
{
  private int _maxHeight = -1;

  public MaxHeightListView(Context context)
  {
    super(context);
  }

  public MaxHeightListView(Context context, AttributeSet attrs)
  {
    super(context, attrs);
    init(context, attrs);
  }

  public MaxHeightListView(Context context, AttributeSet attrs, int defStyle)
  {
    super(context, attrs, defStyle);
    init(context, attrs);
  }

  private void init(Context context, AttributeSet attrs)
  {
    if (attrs != null)
    {
      // Read maxHeight from XML attributes
      for (int i = 0; i < attrs.getAttributeCount(); i++)
      {
        String name = attrs.getAttributeName(i);
        if ("maxHeight".equals(name))
        {
          _maxHeight = attrs.getAttributeIntValue(i, -1);
          break;
        }
      }
    }
  }

  public void setMaxHeight(int maxHeight)
  {
    _maxHeight = maxHeight;
    requestLayout();
  }

  @Override
  protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec)
  {
    if (_maxHeight > 0)
    {
      // Limit height to maxHeight
      int heightMode = View.MeasureSpec.getMode(heightMeasureSpec);
      int heightSize = View.MeasureSpec.getSize(heightMeasureSpec);

      if (heightMode == View.MeasureSpec.UNSPECIFIED || heightSize > _maxHeight)
      {
        heightMeasureSpec = View.MeasureSpec.makeMeasureSpec(_maxHeight, View.MeasureSpec.AT_MOST);
      }
    }

    super.onMeasure(widthMeasureSpec, heightMeasureSpec);
  }
}

File: srcs/juloo.keyboard2/MaxHeightListView.kt
package juloo.keyboard2

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.ListView

/**
 * A scrollable ListView that limits its maximum height.
 * Used for pinned clipboard items - shows max 2 items with internal scrolling.
 */
class MaxHeightListView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : ListView(context, attrs, defStyle) {

    private var maxHeight = -1

    init {
        if (attrs != null) {
            // Read maxHeight from XML attributes
            for (i in 0 until attrs.attributeCount) {
                val name = attrs.getAttributeName(i)
                if ("maxHeight" == name) {
                    maxHeight = attrs.getAttributeIntValue(i, -1)
                    break
                }
            }
        }
    }

    fun setMaxHeight(maxHeight: Int) {
        this.maxHeight = maxHeight
        requestLayout()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var adjustedHeightSpec = heightMeasureSpec

        if (maxHeight > 0) {
            // Limit height to maxHeight
            val heightMode = View.MeasureSpec.getMode(heightMeasureSpec)
            val heightSize = View.MeasureSpec.getSize(heightMeasureSpec)

            if (heightMode == View.MeasureSpec.UNSPECIFIED || heightSize > maxHeight) {
                adjustedHeightSpec = View.MeasureSpec.makeMeasureSpec(maxHeight, View.MeasureSpec.AT_MOST)
            }
        }

        super.onMeasure(widthMeasureSpec, adjustedHeightSpec)

        // Respect minHeight if set
        val minHeight = suggestedMinimumHeight
        if (minHeight > 0) {
            val measuredHeight = measuredHeight
            if (measuredHeight < minHeight) {
                // Enforce minHeight
                setMeasuredDimension(measuredWidth, minHeight)
            }
        }
    }
}
